### PR TITLE
Redirect to shields.io instead of writing our own badges

### PR DESCRIPTION
**Pull request by jyn514**
shields.io has supported docs.rs for a few months now. We don't support
as many features as they do, and they support all features we do, so
just forward requests on to them.

Closes https://github.com/rust-lang/docs.rs/issues/192.

### CODE REVIEW

**Comment by syphar** [comment]
> This is WIP until I figure out how to test it without making network
> requests (I spent way to much time on it already).

@jyn514 can't you just set the redirect policy for `reqwest`? https://docs.rs/reqwest/0.11.3/reqwest/redirect/struct.Policy.html

This would need a new `assert_redirect` method which really only tests the redirect

**Comment by jyn514** [comment]
@syphar I tried that, but it redirected to localhost for some reason: https://github.com/rust-lang/docs.rs/pull/1362/files#diff-e7bb401ce4e1deafed2f3826e6fa900e3f76181906759cbbec3eaf92812381d9R974. I think the localhost override is messing things up: https://github.com/rust-lang/docs.rs/blob/55c7386b9178dc3d33e34508aad29a9b8f4f2639/src/test/mod.rs#L382

**Comment by jyn514** [comment]
Yay, this is finally ready for review :tada:

File: src/web/rustdoc.rs [hunk 1]
```diff
@@ -290,7 +290,7 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
         );
         let url = ctry!(req, Url::parse(&redirect_path));
 
-        Ok(super::redirect(url))
+        Ok(Response::with((status::MovedPermanently, Redirect(url))))
```
Comment (syphar) [inline src/web/rustdoc.rs:293]: why this change? This redirect isn't used in the badge-handler, also it changes the temporary redirect to a permanent one. 

I didn't check the impact of it, but this probably should be a separate PR about redirects with doc pages?

File: src/web/rustdoc.rs [hunk 2]
```diff
@@ -1040,17 +980,43 @@ mod test {
...
+                use reqwest::blocking::Client;
+                use reqwest::redirect::Policy;
+                // avoid making network requests
+                frontend.client = Client::builder().redirect(Policy::none()).build().unwrap();
+            });
+            let mut last_url = "/zstd/badge.svg".to_owned();
+            let mut response = frontend.get(&last_url).send()?;
+            let mut current_url = response.url().clone();
+            // follow redirects until it actually goes out into the internet
+            while !matches!(current_url.host(), Some(Host::Domain(_))) {
```
Comment (syphar) [inline src/web/rustdoc.rs:1001]: why can we have more than one redirect here? shouldn't `/zstd/badge.svg` directly return the redirect to shields.io?

**Review by syphar — CHANGES_REQUESTED** [review]
I tested the endpoint and the redirect seems to work as it should. 
I had to read the `override_frontend` part twice, but it makes sense now. 

Some questions inline.

File: templates/core/about/badges.html [hunk 3]
```diff
@@ -5,51 +5,7 @@
...
-		<tr>
-			<td>
-			Version 0.1.0: <a href="{{ mio_badge | safe }}?version=0.1.0">{{ mio_badge }}?version=0.1.0</a>
-			</td>
-			<td><img src="{{ mio_badge | safe }}?version=0.1.0" alt="mio" /></td>
-		</tr>
-		</tbody>
-	</table>
-	</div>
+	<p>Docs.rs no longer has its own badges. Consider using <a href="https://shields.io">shields.io</a> instead.</p>
```
[the remark this answers is not in the document]
Comment (jyn514) [reply templates/core/about/badges.html:9]: The problem is shields.io doesn't have separate URLs for websites, only for badges. If you go to https://shields.io/ and click "docs.rs" the URL doesn't change.

File: src/web/rustdoc.rs (hunk 1 above)
Comment (jyn514) [reply src/web/rustdoc.rs]: uhhh good point, I think this is left over from trying to get the tests to pass. Reverted now.

File: src/web/rustdoc.rs (hunk 2 above)
Comment (jyn514) [reply src/web/rustdoc.rs:1001]: The test handler redirects from the relative path to an absolute HTTP scheme:
```
[2021-07-06T00:17:45Z INFO  docs_rs::web] Running docs.rs web server on http://127.0.0.1:40949
(/zstd/badge.svg -> http://127.0.0.1:40949/zstd/badge.svg)
(http://127.0.0.1:40949/zstd/badge.svg -> https://img.shields.io/docsrs/zstd/latest)
```

File: templates/core/about/badges.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (syphar) [reply templates/core/about/badges.html:9]: I meant we could just add the template URLs here in the docs, directly the `img.shields.io` one

[the remark this answers is not in the document]
Comment (jyn514) [reply templates/core/about/badges.html:9]: You mean like `https://img.shields.io/docsrs/<crate>/<version>`? That doesn't seem very user friendly - you'd have to type it in without any feedback instead of having the nice ui on shields.io: 
image
and there wouldn't be any indication that there are other options besides the default shape and color.

[the remark this answers is not in the document]
Comment (syphar) [reply templates/core/about/badges.html:9]: I would have loved a direct link to this page to save the user some searching, but checking it I see that it's not possible. 

We only could save one click by pointing to the `build` category.

[the remark this answers is not in the document]
Comment (syphar) [reply templates/core/about/badges.html:9]: I'll leave this fine as it is, except when you want  to change it :)

**Review by syphar — APPROVED** [review]
So, I did another review. Thanks for the answers to my questions @jyn514  :) 

Things I tested: 
- the redirect for latest and specific versions
- the about page
- compared error status codes for docs.rs and shields.io, both return 200 for missing crate or version
- compared caching headers

Things that are changed or are edge-cases that might not work
- some clients might not follow the redirect and not show the badge any more 
- caching headers are different, shields.io seems to have a 30 second cache. 


While the whole test-setup feels more complex than it has to be for this small endpoint, I also don't see an easy way to simplify it that is not too much effort for a change that small.
